Add `woo:validate-zip <path>` command

rcstr opened the pull request:
Part of https://github.com/Automattic/woocommerce.com/issues/20009

### Description
PR adds a basic zip file validation command. Validation is based on:

https://github.com/Automattic/woocommerce.com/pull/20050/files#diff-d1f39d6ef6800717504f5c7dffecf7a012b52659250ba19e7b84f5dbba532c83R120-R126

### Testing instructions
1. Checkout branch
2. Use https://github.com/Automattic/woocommerce.com/pull/20050/files?show-viewed-files=true&file-filters%5B%5D=#diff-839d4fbb14d0686c8ed4cc751e54a9d9a5b586bfabc494cfcd68f0ff53b27ed4 and https://github.com/Automattic/woocommerce.com/pull/20050/files?show-viewed-files=true&file-filters%5B%5D=#diff-36f515fa8eb2124c6844a1ab812d106ab7c677b1973b19e03c721704a518fa39 folders to generate invalid zip files
3. Run `make build`
4. To run the command, use `woo:validate-zip <path>`. Where `<path>` is the path for the zip file, eg `woo:validate-zip /Users/rommel/Projects/woocommerce.com/tests/integration/data/plugins/invalid-file-inside.zip`

### Screenshots
<img width="1204" alt="Screenshot 2024-05-03 at 04 07 09" src="https://github.com/woocommerce/qit-cli/assets/532402/909e155e-ba3d-4a66-aa6b-ef208d6091bc">

rcstr committed:
add dev.md doc to document `detected dubious ownership in repository` error

rcstr committed:
add `\QIT_CLI\Zip` helper

rcstr committed:
add `run:validate-zip <zip>` command

rcstr committed:
init `\QIT_CLI\Zip` tests

rcstr committed:
document running tests

rcstr committed:
extract `\QIT_CLI_Tests\ZipBuilderHelper`

rcstr committed:
update docblock

rcstr committed:
require `\QIT_CLI_Tests\ZipBuilderHelper`

rcstr committed:
fix phpcs warningz

rcstr committed:
extract `\QIT_CLI_Tests\InjectRequestBuilderMockTrait`

rcstr committed:
test valid zip

rcstr committed:
skip `/vendor` and `/node_modules`

rcstr committed:
test `\QIT_CLI\Zip`

rcstr committed:
fix phpcs warnings

rcstr committed:
fix phpstan warnings

Luc45 reviewed (changes requested):
howdy @rcstr , thanks for putting this together!

I'll do a comprehensive code review pass with my impressions, some of them are important, some of them can be considered minor nitpicks, and I'll group all these feedback in an objective, straightforward list:

- First, I would put this command in a new, dedicated namespace called `woo:validate-zip`, where this namespace will be for WooCommerce.com Marketplace-specific operations. This way, we keep the `run` namespace for test types, which are more often than not, independent from the Marketplace. Our vision for the long-term is to open QIT to other extensions as well, such as those on WordPress.org, so keeping Marketplace-specific code containerized helps us achieve that goal.
- Similarly, I'd move the `Zip` class from the parent namespace to a `Woo` namespace.
- I'd rename the `Zip` class to `ZipWooValidator` or similar, to better denote the purpose of the class just by glancing at it.
- Can you please clarify why static is needed on the `Zip` class?
- If we remove statics, we can add the `Zip` class to the constructor of the `RunValidateZipCommand`, leveraging the DI container.
- Just something to keep in mind: This check on the `Zip` class: `$opened = $zip->open( $filepath, ZipArchive::CHECKCONS );` might fail zip files that are created using MacOS Archive Utility. I'm not sure how to deal with this, I guess it depends whether this is a requirement upstream (on the Marketplace) as well. Ultimately, zips created by the MacOS Archive Utility can be installed in WordPress with no problems, as WordPress Core will fallback to pclzip when that ZipArchive checkcons check fails, so should we block it based on that? Debatable. But maybe we could inform the user with a meaningful message?

rcstr committed:
rename command to `woo:validate-zip`

rcstr committed:
update `woo:validate-zip` description and output messages

rcstr committed:
rename `ZIP` to `QIT_CLI\Woo\ZipValidator`. Move to namespace `Woo` and make it a service

rcstr committed:
add message for ZIP files created by `Archive Utility`

rcstr commented:
@Luc45 - thanks for the review. I have handled your points:

- Renamed command to `woo:validate-zip` c9c5b43
- Move `ZIP` to `QIT_CLI\Woo\ZipValidator` class and make it a service 1c68d32
- Added message for _Archive Utility_ possibly created files 6ba3ed8

Luc45 commented on src/src/Woo/ZipValidator.php:
```diff
@@ -0,0 +1,125 @@
+		$opened = $zip->open( $filepath, ZipArchive::CHECKCONS );
+
+		if ( $opened === 21 ) {
+			throw new \RuntimeException( 'Zip file is inconsistent. Zip files generated by the Archive Utility from macOS may be the cause.' );
```
A nitpick, but here we are failing any zips generated by MacOS Archive Utility.

What about instead, we do this?

If it fails the consistency check in a way that matches those zips generated by MacOS Archive Utility, we proceed with the script and show a warning to the user.

```php
	private function open_file( string $filepath ) {
		if ( $opened === 21 ) {
			throw new \UnexpectedValueException( 'Zip file is inconsistent. Zip files generated by the Archive Utility from macOS may be the cause.' );
		}
	}

	public function validate_zip( string $filepath ) {
		try {
			$zip = $this->open_file( $filepath );
		} catch( \UnexpectedValueException $e ) {
			// Let the user know if it fails consistency check because it was generated on Mac, but proceed.
			// Any other exception (such as RuntimeException) will not be caught and will interrupt the script.
			\QIT_CLI\App::make( \QIT_CLI\IO\Output::class )->writeln( sprintf( '<comment>%s</comment>', $e->getMessage() ) );
		}
	}
```

rcstr replied:
Implemented in 6ba3ed8...060fa17

<img width="879" alt="Screenshot 2024-05-09 at 15 10 49" src="https://github.com/woocommerce/qit-cli/assets/532402/b5951ff7-1390-464f-a971-cb157228d4f5">

rcstr committed:
throw `UnexpectedValueException` is file inconsistent

rcstr committed:
catch `UnexpectedValueException`

rcstr committed:
test an inconsistent zip file

zhongruige commented on docs/dev.md:
```diff
@@ -0,0 +1,15 @@
+### `detected dubious ownership in repository` error during `make build
```
Thanks for adding in this doc update! Only a minor adjustment:

```suggestion
### `detected dubious ownership in repository` error during `make build`
```

zhongruige commented on src/src/Commands/WooValidateZipCommand.php:
```diff
@@ -0,0 +1,48 @@
+
+			return Command::FAILURE;
+		} catch ( \Exception $e ) {
+			$output->writeln( sprintf( '<error>An error occurred while validating the ZIP file. Error: %s</error>', $e->getMessage() ) );
```
Minor wording change suggestion:

```suggestion
			$output->writeln( sprintf( '<error>An error occurred while validating the provided file. Error: %s</error>', $e->getMessage() ) );
```

zhongruige commented on src/src/Woo/ZipValidator.php:
```diff
@@ -0,0 +1,125 @@
+				}
+
+				if ( $this->is_file_invalid( $info['name'] ) ) {
+					throw new \RuntimeException( sprintf( 'Invalid (%s) file/folder inside zip file', $info['name'] ), 400 );
```
Minor wording change:

```suggestion
					throw new \RuntimeException( sprintf( 'Invalid (%s) file/folder inside the provided ZIP file', $info['name'] ), 400 );
```

zhongruige commented on src/src/Woo/ZipValidator.php:
```diff
@@ -0,0 +1,125 @@
+				$info = $zip->statIndex( $i );
+
+				if ( ! $info ) {
+					throw new \RuntimeException( 'Cannot read zip index.', 400 );
```
We use both `zip` and `ZIP` interchangeably, so we may want to settle on one. I think `ZIP` is the general standard, so maybe we can settle on that?

rcstr replied:
updated `zip/Zip` to `ZIP` in c181a53

zhongruige reviewed (commented):
So far this has been testing great for me with both valid and invalid zip files, nice work. I added a few comments around some suggested wording changes, let me know your thoughts!

rcstr committed:
update error wording

Co-authored-by: Greg <71906536+zhongruige@users.noreply.github.com>

rcstr committed:
close comment inlin-block

Co-authored-by: Greg <71906536+zhongruige@users.noreply.github.com>

rcstr committed:
update exception message

Co-authored-by: Greg <71906536+zhongruige@users.noreply.github.com>

rcstr committed:
make `ZIP` uppercase

Luc45 committed:
Merge branch 'trunk' into 24-04/implement-incompatible-archive-check

Luc45 committed:
Test tweaks

Luc45 committed:
Edit dev.md

Luc45 committed:
Remove unnecessary trait, manual require, move to dedicated helpers dir

Luc45 committed:
Tweak

zhongruige reviewed (approved):
Ran through the changes again after the merge conflicts were resolved and everything tested great again for me:

<img width="911" alt="Screenshot 2024-05-13 at 4 06 52 PM" src="https://github.com/woocommerce/qit-cli/assets/71906536/554bb0c4-0e25-4332-82ed-2122932540f3">

Nice work on this @rcstr! After @Luc45 reviews the changes he had originally requested are resolved, we should be all set.

Once this is done, I believe we can get this [added to our documentation](https://qit.woo.com/docs/category/managed-tests), I think under "Managed Tests" might make the most sense, and we can mention the files and folder structures this checks for to provide more context for the user, then in a follow up task, add a link to the docs in the results. Let me know your thoughts!
